src/js/tooltip.js

```diff
@@ -0,0 +1,420 @@
+				rect.top = this._getTopFor('middle');
+				break;
+
+			case 'left':
+				rect.left = this.target.left - (width + Tooltip.arrowDepth);
+				rect.top = this._getTopFor('middle');
+				break;
+		};
+
+		rect.right = rect.left + width;
+		rect.bottom = rect.top + height;
+
+		return rect;
+	}
+
+	_getLeftFor(alignment) {
+		if (alignment === "middle") {
+			return this.target.centrePoint.x - this.width()/2;
+		} else if (alignment === "left") {
+			return this.target.left;
+		} else if (alignment === "right") {
+			return this.target.right - this.width();
+		}
+	};
+
+	_getTopFor(alignment) {
+		if (alignment === "middle") {
+			return this.target.centrePoint.y - this.height()/2;
+		} else if (alignment === "top") {
+			return this.target.top;
+		} else if (alignment === "bottom") {
+			return this.target.bottom - this.height();
+		}
+	}
+
+	_setArrow() {
+		const arrowClassRoot = 'o-tooltip--arrow-';
+		const alignmentClassRoot = 'o-tooltip-arrow--align-';
+
+		let classesToRemove = ["o-tooltip--arrow-left",
+		"o-tooltip--arrow-right",
+		"o-tooltip--arrow-above",
+		"o-tooltip--arrow-below",
+		"o-tooltip-arrow--align-top",
+		"o-tooltip-arrow--align-bottom",
+		"o-tooltip-arrow--align-left",
+		"o-tooltip-arrow--align-right"];
+
+		this.tooltipEl.classList.remove(...classesToRemove);
+		this.tooltipEl.classList.add(arrowClassRoot + Tooltip.positionToArrowPositionMap[this.tooltipPosition]);
+		this.tooltipEl.classList.add(alignmentClassRoot + this.tooltipAlignment);
+	};
+
+	_drawTooltip(rect) {
+		this.tooltipEl.style.top = rect.top + 'px';
+		this.tooltipEl.style.left = rect.left + 'px';
+	};
+
+	static _isOutOfBounds(point, axis) {
+		if (point < 0) {
+			return true;
+		}
+		if (axis === 'y' && point > document.body.offsetHeight) {
+			return true;
+		} else if (axis === 'x' && point > document.body.offsetWidth) {
+			return true;
+		}
+		return false;
+	}
+
+	static _flipOrientation(orientation) {
+		const flip = {
+			"above": "below",
+			"below": "above",
+			"left": "right",
+			"right": "left"
+		};
+
+		return flip[orientation];
+	}
+
+	static throwError(message) {
+		throw new Error('"o-tooltip error": '+ message);
+	};
+
+	static init (rootEl, opts) {
+		if (!rootEl) {
+			rootEl = document.body;
+		}
+
+		// If the rootEl isnt an HTMLElement, treat it as a selector
+		if (!(rootEl instanceof HTMLElement)) {
+			rootEl = document.querySelector(rootEl);
+		}
+
+		// If the rootEl is an HTMLElement (ie it was found in the document anywhere)
+		// AND the rootEl has the data-o-component=o-tooltip then initialise just 1 tooltip (this one)
+		if (rootEl instanceof HTMLElement && /\bo-tooltip\b/.test(rootEl.getAttribute('data-o-component'))) {
+			return new Tooltip(rootEl, opts);
+		}
+
+		// If the rootEl wasn't itself a tooltip, then find ALL of the child things that have the data-o-component=oTooltip set
+		return [].map.call(rootEl.querySelectorAll('[data-o-component="o-tooltip"]'), rootEl => new Tooltip(rootEl, opts));
```

Comment:
This can be done using `Array.from` instead, which should be simpler to understand.

`Array.from(rootEl.querySelectorAll('[data-o-component="o-tooltip"]'), rootEl => new Tooltip(rootEl, opts))`